Legacy reports: Decouple the dependency between the legacy reports and the WooCommerce Status widget on the dashboard page

eason9487 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This is one of the preparations for gradually retiring the WooCommerce legacy reports.

It focuses on decoupling the dependency between the legacy reports and the WooCommerce Status widget on the dashboard page.

- Reverse the dependency on drawing sparklines in the legacy reports from the WooCommerce Status widget on the dashboard page.
- Decouple the dependency between the legacy reports and the WooCommerce Status widget on the dashboard page.
- Break down the method of forming HTML for the sparkline in Dashboard widget to avoid directly outputting HTML via a filtered callback.

In addition, It adds the new file `woocommerce-legacy-reports.php` to simulate the loading plugin for the legacy reports. In this way, we can move other dependencies of legacy reports to the new file in subsequent PRs.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### View the WooCommerce Status widget before this PR

1. Checkout the current `trunk`, and run `pnpm install` and `pnpm build`
2. Place some orders if there is no order in this month
3. Go to the admin page of *WooCommerce > Settings > Advanced > Features*. Disable **WooCommerce Analytics**
   - You may need to sync order or enable compatibility mode for Order data storage
4. Go to the admin page of *Dashboard*
5. View the **WooCommerce Status** widget
    ![1](https://github.com/user-attachments/assets/d4146fdc-4183-4f77-8775-1e340abdff6b)
   - Two sparklines
   - Four hyperlinks to the pages of legacy reports
6. Enable **WooCommerce Analytics** and refresh the Dashboard page
7. View the **WooCommerce Status** widget
   - Two sparklines (might be a bit difference from the ones in the above step because their queries are not exactly the same)
   - Four hyperlinks to the pages of Analytics

#### Check the WooCommerce Status widget after this PR changes

1. Checkout this PR, and run `pnpm install` and `pnpm build`
2. Run `pnpm install` and `pnpm build`
3. Repeat steps 3-7 in the above section of testing instructions
4. Check if the sparklines and hyperlinks are the same after this PR changes

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

eason9487 committed:
Reverse the dependency on drawing sparklines in the legacy reports from the WooCommerce Status widget on the dashboard page.

eason9487 committed:
Add a new class to simulate the loading plugin for the legacy reports.

eason9487 committed:
Decouple the dependency between the legacy reports and the WooCommerce Status widget on the dashboard page.

eason9487 committed:
Break down the method of forming HTML for the sparkline in Dashboard widget to avoid directly outputting HTML via a filtered callback.

eason9487 committed:
Update PHP unit tests for WC_Tests_Admin_Dashboard.

eason9487 committed:
Add changelog.

mikkamp commented on plugins/woocommerce/includes/admin/class-wc-admin-dashboard.php:
```diff
@@ -309,13 +326,8 @@ private function status_widget_stock_rows( $is_wc_admin_disabled ) {
-				$outofstock_link = 'admin.php?page=wc-admin&type=outofstock&path=%2Fanalytics%2Fstock';
-			}
+			$lowstock_url     = $lowstock_link ? admin_url( $lowstock_link ) : '#';
+			$outofstock_url   = $outofstock_link ? admin_url( $outofstock_link ) : '#';
```
If we don't have a link available, wouldn't it be better to output it as regular text instead of linking it to the current page? Or does that become too complex as we'd need to adjust the current CSS to match the layout?

eason9487 replied:
Yes, to present a similar UI without the `a` tag will require some CSS adjustments.

However, the reason I didn't do that is not about complexity but rather about the fact that it doesn't expect to get invalid links here, so I just did a low level of fault tolerance and didn't implement a separate UI for invalid links.

mikkamp commented on plugins/woocommerce/includes/admin/class-wc-admin-dashboard.php:
```diff
@@ -344,7 +356,7 @@ private function status_widget_stock_rows( $is_wc_admin_disabled ) {
 			?>
 			<li class="low-in-stock">
-			<a href="<?php echo esc_url( admin_url( $lowstock_link ) ); ?>">
+			<a href="<?php echo esc_url( $lowstock_url ); ?>">
```
The indentation of the `a` tag is not aligned with the closing tag. It's not really caused by this PR but we could align that here.

eason9487 replied:
Adjusted in https://github.com/woocommerce/woocommerce/pull/51566/commits/22163608935c788b3cb041bffe869c4dba14d7eb.

mikkamp commented on plugins/woocommerce/includes/admin/class-wc-admin-dashboard.php:
```diff
@@ -211,9 +226,10 @@ public function status_widget() {
 			}
 
+			$reports = $status_widget_reports['legacy_report'];
 			do_action( 'woocommerce_after_dashboard_status_widget', $reports );
```
`legacy_reports` is only being set to an instance of `WC_Admin_Report` when the filter `woocommerce_dashboard_status_widget_reports` is called. But when analytics is enabled it remains null (because the filter isn't called). In theory that's breaking behaviour if anyone has a code snippet/plugin where they were using that parameter. Is there a better way to handle this gracefully?

I can't find any code snippets where the parameter is actually used, but that doesn't mean there aren't sites / plugins where it is being used.

For example this code snippet can cause a fatal error only when analytics is enabled:
```php
add_action(
	'woocommerce_after_dashboard_status_widget',
	function ( $reports ) {
		$reports->output_report();
	}
);
```

eason9487 replied:
I searched all the repos under GitHub woocommerce organization during development and found that none of the locations that use this action use the parameter, so I originally wanted to gradually make it optional.

However, I can't be sure if there are third-party plugins use it, so it's still better to keep it unchanged in this PR. Adjusted in https://github.com/woocommerce/woocommerce/pull/51566/commits/62e9cf0c346ce302eda87dee1dcc1fea39e6cb85

mikkamp commented on plugins/woocommerce/includes/admin/class-wc-admin-dashboard.php:
```diff
@@ -526,20 +538,14 @@ private function get_wc_admin_performance_data() {
-			if ( $is_wc_admin_disabled ) {
-				return $reports->sales_sparkline( $id, $days, $type );
-			}
+		private function get_sales_sparkline( $id = '', $days, $type = 'sales' ) {
```
Required parameters shouldn't be added after an optional parameter since PHP 8. So this gives me the following deprecation message:

```
[20-Sep-2024 14:52:19 UTC] PHP Deprecated:  Optional parameter $id declared before required parameter $days is implicitly treated as a required parameter in woocommerce/includes/admin/class-wc-admin-dashboard.php on line 548
```

eason9487 replied:
Thanks for catching this! Fixed in https://github.com/woocommerce/woocommerce/pull/51566/commits/e153dce12e12946d6e9dff899f4781520b8f4de1.

mikkamp commented on plugins/woocommerce/includes/admin/reports/class-wc-admin-report.php:
```diff
@@ -611,6 +611,28 @@ public function sales_sparkline( $id = '', $days = 7, $type = 'sales' ) {
+	 *
+	 * @param  int    $id ID of the product to show. Blank to get all orders.
+	 * @param  int    $days Days of stats to get. Default to 7 days.
+	 * @param  string $type Type of sparkline to get. Ignored if ID is not set.
```
Type is not really ignored here if ID is not set, as it still influences which tooltip text is shown.

eason9487 replied:
Adjusted in https://github.com/woocommerce/woocommerce/pull/51566/commits/e8f8e772b995974f372868bd42c2ba922014c8f3.

mikkamp reviewed (commented):
Thanks for the changes here. When testing the code continues to work as it previously did. However I came across a few quirks when going through the code changes.

eason9487 committed:
Ensure the backward compatibility for the `woocommerce_after_dashboard_status_widget ` action.

Address: https://github.com/woocommerce/woocommerce/pull/51566#discussion_r1768751063

eason9487 committed:
Fix the incorrect indentation of a modified line of codes in the `WC_Admin_Dashboard` class.

Address: https://github.com/woocommerce/woocommerce/pull/51566#discussion_r1768690182

eason9487 committed:
Fix a PHP 8 deprecation message due to a required parameter after an optional parameter in the `WC_Admin_Dashboard` class.

Address: https://github.com/woocommerce/woocommerce/pull/51566#discussion_r1768776896

eason9487 committed:
Tweak the inaccurate PHPDoc for the `get_sales_sparkline ` method in the `WC_Admin_Report` class.

Address: https://github.com/woocommerce/woocommerce/pull/51566#discussion_r1768804777

eason9487 commented:
Hi @mikkamp, thanks for the code review!

I have made some changes. Could you help with a new round of code reviews?

mikkamp reviewed (approved):
Thanks for all the additional changes. They look good to me, it continues to run as intended.
